kakmerge

```diff
@@ -162,6 +163,41 @@ do_create_layout_tmux() {
         select-pane -D
 }
 
+do_create_layout_kitty() {
+    kak_session="${1}"
+
+    kitty @ goto-layout fat
+    [ $? -ne 0 ] && exit 1  # kitty prints a nice message if it doesn't support remote control
+
+    sleep 10 &
+    wait_local=$!
+    sleep 10 &
+    wait_base=$!
+    sleep 10 &
```

Comment:
`10` is arbitrary, here I basically allow at most 10 seconds for kakoune to initialize, not to block `wait` below indefinitely.

Reply:
Since the calls to `kitty` are synchronous, if they (and the `kak` process in them) take more than 10 seconds, then `wait` will not block, and the synchronization will fail.

I have `allow_remote_control yes` temporarily set in my configuration.

Reply:
Also why use `exec %sh{ … }`, instead of `nop %sh{ … }`?

Reply:
If they take more than 10 seconds, I assume something is wrong, it sounds wrong if kakmerge takes more than 10 seconds to launch, don't you agree?

Reply:
> Also why use exec %sh{ … }, instead of nop %sh{ … }?

Still learning the syntax, will adapt 🙂

Reply:
> It is required to enable remote control in kitty

This message in your stack trace suggests that your exception happens on setting the "fat" layout, before the code reaches synchronization stage.

Reply:
Reported the issue as #1169@kitty.

I have an alternative implementation that I cannot test because of the above issue:

```diff
diff --git a/kakmerge b/kakmerge                                                 
index ed03420..a25d812 100755                                                    
--- a/kakmerge                                                                   
+++ b/kakmerge                                                                   
@@ -169,33 +169,35 @@ do_create_layout_kitty() {                                 
     kitty @ goto-layout fat                                                     
     [ $? -ne 0 ] && log_fatal "It is required to enable remote control in kitty"
                                                                                 
-    sleep 10 &                                                                  
-    wait_local=$!                                                               
-    sleep 10 &                                                                  
-    wait_base=$!                                                                
-    sleep 10 &                                                                  
-    wait_remote=$!                                                              
+    pipe_local="${TMPDIR:-/tmp}/kakmerge-local-${kak_session}"                  
+    mkfifo "${pipe_local}"                                                      
+    pipe_remote="${TMPDIR:-/tmp}/kakmerge-remote${kak_session}"                 
+    mkfifo "${pipe_base}"                                                       
+    pipe_remote="${TMPDIR:-/tmp}/kakmerge-remote-${kak_session}"                
+    mkfifo "${pipe_remote}"                                                     
                                                                                 
     kitty @ new-window kak -c "${kak_session}" \                                
                        -e '                                                     
                        rename-client local                                      
                        rename-session %{'"${kak_session}"'}                     
-                       nop %sh{ kill '"${wait_local}"' }                        
+                       nop %sh{ echo > "'"${pipe_local}"'" }                    
                        ' >/dev/null                                             
     kitty @ new-window kak -c "${kak_session}" \                                
                        -e '                                                     
                        rename-client base                                       
                        rename-session %{'"${kak_session}"'}                     
-                       nop %sh{ kill '"${wait_base}"' }                         
+                       nop %sh{ echo > "'"${pipe_base}"'" }                     
                        ' >/dev/null                                             
     kitty @ new-window kak -c "${kak_session}" \                                
                        -e '                                                     
                        rename-client remote                                     
                        rename-session %{'"${kak_session}"'}                     
-                       nop %sh{ kill '"${wait_remote}"' }                       
+                       nop %sh{ echo > "'"${pipe_remote}"'" }                   
                        ' >/dev/null                                             
     kitty @ focus-window --match id:"${KITTY_WINDOW_ID}"                        
-    wait "${wait_local}" "${wait_base}" "${wait_remote}"                        
+                                                                                
+    cat "${pipe_local}" "${pipe_base}" "${pipe_remote}" >/dev/null              
+    rm -f "${pipe_local}" "${pipe_base}" "${pipe_remote}"                       
 }                                                                               
                                                                                 
 cmd_merge() {                                                                   
```

I think it's more idiomatic to use pipes as synchronization mechanisms, but I looked at the [POSIX page for `sleep`](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/sleep.html) and it states:

>_sleep_ is required by this volume of POSIX.1-2017 to deal with _time_ requests of up to 2147483647 seconds

Recent implementations of `sleep` allow `infinity` for parameter, but a POSIX equivalent to that would be simply `infinity=2147483647; sleep $infinity` (as per the above quote), which would make your implementation less wobbly.

Also `wait` prints messages in my shell, so a `>/dev/null` would be needed.

Reply:
Your approach with pipes works equally well, the only difference is that in case of issues terminal is left in a broken state.

In other words, if I just remove `nop %sh{ kill '"${wait_local}"' }` line from this branch, after 10 seconds I'll get an error and that's it.
If I apply your patch and then remove `nop %sh{ echo > "'"${pipe_local}"'" }` to simulate a failure, it will hang forever, and if I cancel with Ctrl+C, I'll have kitty in weird state with 3 frozen small windows.

I prefer my approach, but I wouldn't object if you want me to use your patch instead. Do you want me to?

> I have an alternative implementation that I cannot test because of the above issue:

Can you temporarily disable tmux for testing? I can already tell you what Kovid will do with your issue, he hates tmux so he will close it with a comment "not interested in debugging tmux" 😛

Reply:
Like I said, we currently have no way of knowing whether kak has successfully loaded its configuration or executed a command, so we're assuming that this part will run fine.

Change your code to make `sleep` use `infinity=2147483647`, redirect the output of `wait` to /dev/null, then I think we'll be done.

Reply:
As you wish - pushed